Elastic-plastic bond style

jtclemm opened the pull request:
**Summary**

Includes a new elastic-perfectly-plastic BPM bond style, a new approximation for the energy of the bpm/spring bond style, two examples that demonstrate plastic deformation and varying Poisson's ratio (incompressible -> auxetic), and a few other minor package updates. 

Videos of examples below for fun.

**Author(s)**

Joel Clemmer (SNL)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

NA

**Implementation Notes**

NA

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

![elastic-plastic-demo1.gif](https://github.com/user-attachments/assets/68232b47-dbab-4e68-8fa8-d11be9a3a10e)
![elastic-plastic-demo2.gif](https://github.com/user-attachments/assets/b0f44e37-caf0-4d3f-8e55-5f7a4c976137)

jtclemm committed:
adding bpm energy proxy

jtclemm committed:
Initial plastic BPM bond style

jtclemm committed:
BPM plastic citation info

jtclemm committed:
Drafting bpm/plastic doc page

jtclemm committed:
Adding auxetic/incompressible bpm demo

jtclemm committed:
Adding bpm/plastic demo, fixing mistake in poisson example

jtclemm committed:
Tweaking BPM plastic example parameters

jtclemm committed:
fix doc spacing for bpm/spring

jtclemm committed:
fixing typo

jtclemm committed:
Adding references to bpplastic bpm bond style

jtclemm committed:
Adding BPM contribution statements

jtclemm committed:
Merge branch 'develop' into BPM

jibril-b-coulibaly commented on doc/src/bond_bpm_spring_plastic.rst:
```diff
@@ -0,0 +1,183 @@
+
+   bond_style bpm/spring/plastic keyword value attribute1 attribute2 ...
+
+* optional keyword = *overlay/pair* or *store/local* or *smooth* or *break*
```
The keywords list do no include `normalize`.
This is also the case for the `bpm/spring` documentation

jibril-b-coulibaly commented on doc/src/bond_bpm_spring_plastic.rst:
```diff
@@ -0,0 +1,183 @@
+below. If the bond stretches beyond a strain of :math:`\epsilon_p`, it will
+plastically activate and :math:`r_\mathrm{eq}` will evolve to ensure
+:math:`|(r-r_\mathrm{eq})/r_\mathrm{eq}|` never exceeds :math:`\epsilon_p`.
+Therefore, in pure tension, the force initially grows elastically before
```
The equation above has absolute values. If plastic plateau occurs also in compression it is worth rephrasing this sentence, or adding a new one for clarity

jibril-b-coulibaly commented on doc/src/bond_bpm_spring_plastic.rst:
```diff
@@ -0,0 +1,183 @@
+The single() function also calculates two extra bond quantities, the initial
+distance :math:`r_0` and the current equilbrium length :math:`r_eq`. These extra
+quantities can be accessed by the :doc:`compute bond/local <compute_bond_local>`
+command as *b1* and *b2*\ .
```
add respectively? To be explicit that b1 is r_0 and b2 r_eq

jibril-b-coulibaly commented on src/BPM/bond_bpm_spring_plastic.cpp:
```diff
@@ -0,0 +1,477 @@
+      type = bond_type[i][m];
+
+      //Skip if bond was turned off
+      if (type < 0) continue;
```
Could this have any issue if for some reason a bond is initialized broken with type ==0 ?

jibril-b-coulibaly commented on src/BPM/pair_bpm_spring.cpp:
```diff
@@ -378,8 +386,13 @@ double PairBPMSpring::single(int i, int j, int itype, int jtype, double rsq, dou
   fpair = -k[itype][jtype] * dr;
-  if (anharmonic_flag)
-    fpair += -ka[itype][jtype] * dr * dr * dr;
+  double energy = -0.5 * fpair * (r - cut[itype][jtype]);
```
Consider replacing `(r - cut[itype][jtype])` by `dr`

jibril-b-coulibaly reviewed (commented):
Everything looks good to me. I have minor form comments and a question about type < 0 that you will find in specific comments

jtclemm committed:
Fixing mistakes & clarifying BPM files

jtclemm commented:
> Everything looks good to me. I have minor form comments and a question about type < 0 that you will find in specific comments

Thanks! Good catches. I fixed all the issues and also changed all broken bond tests to `type <= 0`. It was actually inconsistent with `<` and `<=` throughout the package, not just at that one spot. I think it's best to avoid bond type zero too.

sjplimp commented:
@jtclemm I cannot view the 2 movies (top of this page).  Can you send me the MP4 files separately?

@akohlmey If these 2 videos are small, can we include them in the doc page for this new command?  Is there a standard manual format for video files?  I imagine Joel has others for the BPM package.  

Joel, if we don't include things like this in the doc pages, we should put them all in a "BPM" location on the LAMMPS website.  And then point all your doc pages to the website link.  All of this is to attract users to BPM.

sjplimp commented:
Just found a link to this short video on the pair_mesodpd.rst doc page
JPG/examples_mdpd.gif
I don't see any other video files in that dir

akohlmey commented:
> @jtclemm I cannot view the 2 movies (top of this page). Can you send me the MP4 files separately?

I've replaced them with animated GIFs and now they display.

> @akohlmey If these 2 videos are small, can we include them in the doc page for this new command?

Movies are almost always too large. As MPEG4 movie they are about 2MB as animated GIF they are 6MB and 8MB, respectively.

Outside of animated GIFs (which will just show a static image otherwise) there is no standard way I know to have movies included in the manual consistently. I would put those (as MPEG4) on the webserver.

akohlmey commented:
> JPG/examples_mdpd.gif
> I don't see any other video files in that dir

I suspect it could sneak in because nobody suspected it to be an animated GIF.

jtclemm commented:
> @jtclemm I cannot view the 2 movies (top of this page). Can you send me the MP4 files separately?
> 
> @akohlmey If these 2 videos are small, can we include them in the doc page for this new command? Is there a standard manual format for video files? I imagine Joel has others for the BPM package.
> 
> Joel, if we don't include things like this in the doc pages, we should put them all in a "BPM" location on the LAMMPS website. And then point all your doc pages to the website link. All of this is to attract users to BPM.

Missed replying to this. I hadn't created the videos with an intention of adding to the documentation, but if we want to add them to the LAMMPS website I can modify the documentation and include a link. 

I wouldn't want to render the videos as animated gifs because the size.
